### ISSUE TITLE
OpenMM 7.5 release plan

### ISSUE THREAD

**Issue by jaimergp** [openmm/openmm#2905]
Hi!

Peter, could you nominate a commit to start planning the release of v7.5? I want to base all the testing for the Conda Forge migration using something as close as possible to the final release. At least I need the Python 3.9 fixes (#2888) and the PowerPC / ARM fixes (#2499 #2781, maybe others?).

Thanks!

**Comment by peastman** [comment]
Use the current head of the master branch.  That's 2c36e9b34d046f09be4887ff60e7fdac02cf3cdb.  The final release will be something very close to that.

**Comment by peastman** [comment]
Now that the build system is working (thanks to amazing work by @jaimergp!), I think we're just about ready to build the release candidate.  Are there any final changes we should make first?

**Comment by peastman** [comment]
One change we need is to update the installation instructions in the manual.  What will be the commands for installing it, including handling of different CUDA versions?

We also need to create builds of PDBFixer and OpenMM-Setup and update the installation instructions for them.  Those should be easy since they're just Python.

**Comment by jaimergp** [comment]
> What will be the commands for installing it, including handling of different CUDA versions?

I'll add PRs for that.

`pdbfixer` and `openmm-setup` can go to conda-forge once `openmm` is in `main` too. I can have the packages submitted beforehand to have minimal delays.

**Comment by peastman** [comment]
Great, thanks!

**Comment by jaimergp** [comment]
> I can have the packages submitted beforehand to have minimal delays.

All the remaining `omnia` packages are being submitted in this PR: https://github.com/conda-forge/staged-recipes/pull/13234

This won't be merged until OpenMM goes `main`, which will happen once 7.5 is released. In the meantime, we can test the packages through the [`linux` artifacts](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=240554&view=artifacts&type=publishedArtifacts) because most are just `noarch`.

**Comment by peastman** [comment]
Perfect.  That means we should finalize the code revisions for all packages.  @jchodera your thoughts on https://github.com/openmm/openmm-setup/pull/20 would be much appreciated.

**Comment by peastman** [comment]
I think we're now ready to build the release candidate?

I'll make final changes to PDBFixer and OpenMM-Setup so we can build packages for them at the same time as the official OpenMM release, probably sometime next week.

**Comment by jchodera** [comment]
@peastman: Since these packages depend on OpenMM, we can't build them at the same time. Instead, we should get OpenMM out to the conda-forge channel and then worry about the dependencies. But nothing can happen until then, I don't think.

**Comment by peastman** [comment]
Oops - in testing the PDBFixer updates, I just discovered another OpenMM bug that needs to be fixed.  PR coming in a moment.

**Comment by peastman** [comment]
Ok, I merged the fix.  Now we can continue!

**Comment by jaimergp** [comment]
PDBFixer and OpenMM-Setup are included in https://github.com/conda-forge/staged-recipes/pull/13234.

Since these _only_ depend on OpenMM, I can probably split them to a separate PR so it's easier for conda-forge core to review them.

**Comment by jaimergp** [comment]
> Ok, I merged the fix. Now we can continue!

Can we nominate a new hash to work as RC?

**Comment by peastman** [comment]
83e921314a32606bef96311f5e1f91db74021e3c.  That's the current head of the master branch.

**Comment by peastman** [comment]
Should we also include #2935?  It fixes an error in a new feature that was added in 7.5.  The risk is hopefully fairly low, since if you don't enable periodic exceptions, it shouldn't affect anything.  But of course, nothing is without risk.

**Comment by jchodera** [comment]
Seems like it makes sense to not ship a release with known bugs if we can avoid it.

**Comment by peastman** [comment]
Ok, merged.  The head of the master branch is now b49b82efb5a253a7c891ca084b3370e181de2ea3.

**Comment by jaimergp** [comment]
Looks like the last builds at conda-forge are fully operational. Shall we nominate those as the `rc`?

Instructions to install would be:

```bash
conda create -n openmm-750-rc -c conda-forge/label/test -c conda-forge openmm 
conda activate openmm-750-rc
python -m simtk.testInstallation
```

Or with a specific cudatoolkit:

```bash
conda create -n openmm-750-rc -c conda-forge/label/test -c conda-forge openmm cudatoolkit=10.0
conda activate openmm-750-rc
python -m simtk.testInstallation
```

**Comment by jchodera** [comment]
@peastman : I suggest we do the following for this release process:
* Cut a "prerelease" 7.5.0rc1 release on GitHub that will contain all the information about how to download, test, and report issues. Crucially, unlike the SimTK page, @jaimergp and I will both be able to update the release notes. 
* We share that release link on twitter and other channels and ask people to test
* If all goes well, @peastman can cut an official 7.5.0 release on Mon and @jaimergp can update the conda package on Tue morning Europe time

**Comment by peastman** [comment]
Awesome!  The final thing we need to do is a minimal test following the procedure at https://github.com/openmm/openmm/wiki/Release-Candidates.  That means installing on Linux, Windows, and Mac, and for each one verify that 1) all expected platforms work, 2) the version number is correct, and 3) the git revision is correct.  Once that's done I can post to the forum.

> If all goes well, @peastman can cut an official 7.5.0 release on Mon and @jaimergp can update the conda package on Tue morning Europe time

We usually wait a full week after a release candidate.  If we get really a lot of people testing it by Monday we could accelerate it, but that might be pushing it a bit.

**Comment by peastman** [comment]
The Linux build lists

```
Git Revision: c54f33c7c63d0ff58dd90aab52e5ec10e18f74f8
```

That isn't the correct revision.  It's supposed to be b49b82e.

The Mac build lists

```
Git Revision: Unknown
```

**Comment by jaimergp** [comment]
Yes, I am observing the same behavior in my tests.

I am obtaining the source tarballs [using the `/archive` endpoint in GitHub](https://github.com/conda-forge/openmm-feedstock/blob/master/recipe/meta.yaml#L11). Is that supposed to work with the `git_revision` parsing? The revision in these tarballs might not obtainable as expected...

**Comment by jaimergp** [comment]
I am going to try and patch the `version.py` file in the conda recipe. If that doesn't work, we can also try to clone the repository, but that doesn't work as nicely for reproducible builds...

**Comment by peastman** [comment]
> I am obtaining the source tarballs using the /archive endpoint in GitHub. Is that supposed to work with the git_revision parsing?

No idea.  In the past we've specified the revision directly:

https://github.com/omnia-md/conda-recipes-cuda/blob/3798691b217ff9b49a53a1c453846d8b20df92d4/openmm/meta.yaml#L5-L7

**Comment by jchodera** [comment]
@jaimergp : Would it help if we cut a 7.5.0rc1 release and used that for the conda-forge build?

> Once that's done I can post to the forum.

The goal is to collect all the documentation for installing and using it---and any FAQs and answers---in the release notes that we post on GitHub. You can announce this in the forum and link to it, but it is important we do not make the forum the primary source of release information that everything points to.

**Comment by peastman** [comment]
We have a standard procedure for this: https://github.com/openmm/openmm/wiki/Release-Candidates.  Release notes, documentation, FAQs, etc. go with the release, not the rc.

Nothing else needs to point to the forum post, but I'm still going to include the installation instructions in the post.  If we want people to test the release candidate we should make it as easy as possible.  Besides, I'm going to discuss how things have changed since the beta, which will include examples of different ways of installing.

**Comment by peastman** [comment]
Looks like the builds are done, so I'm starting on testing.

**Comment by peastman** [comment]
Everything looks good!  I can't test CUDA on Windows since I don't have a Windows box with an NVIDIA GPU, but that will have to do.

**Comment by jaimergp** [comment]
I'll do the Windows testing tomorrow!

**Comment by jchodera** [comment]
> We have a standard procedure for this: https://github.com/openmm/openmm/wiki/Release-Candidates. Release notes, documentation, FAQs, etc. go with the release, not the rc.

We should revisit that process, especially because conda-forge requires a different process than this for ensuring version numbers are correct.

**Comment by jaimergp** [comment]
I ran these scripts to test the versions locally:

<details>

<summary>Unix</summary>

```bash
#!/bin/bash

set -eo pipefail

revision="b49b82efb5a253a7c891ca084b3370e181de2ea3"
version="7.5"

for cudaver in 9.2 10.0 10.1 10.2 11.0; do
    echo "--- Installing OpenMM for CUDA $cudaver ---"
    conda create -yq -n openmm-cf -c conda-forge/label/test -c conda-forge openmm cudatoolkit=$cudaver
    conda activate openmm-cf
    python -m simtk.testInstallation
    python -c "from simtk.openmm import Platform; v = Platform.getOpenMMVersion(); assert \"${version}\" in (v, v+'.0'), v + \"!=${version}\""
    python -c "from simtk.openmm.version import git_revision; r = git_revision; assert r == \"${revision}\", r + \"!=${revision}\""
    conda deactivate
    echo "--- Done! ---"
done
```

</details>



<details>

<summary>Windows</summary>

```batch
:: Run with `cmd /c this_script.bat`

@echo off

set "revision=b49b82efb5a253a7c891ca084b3370e181de2ea3"
set "version=7.5"

for %%G in (10.0,10.1,10.2,11.0) do (
    echo "--- Installing OpenMM for CUDA %%G ---"
    conda create -yq -n openmm-cf -c conda-forge/label/test -c conda-forge openmm cudatoolkit=%%G || goto :error
    conda activate openmm-cf || goto :error
    python -m simtk.testInstallation || goto :error
    python -c "from simtk.openmm import Platform; v = Platform.getOpenMMVersion(); assert '%version%' in (v, v+'.0'), v + '!=%version%'" || goto :error
    python -c "from simtk.openmm.version import git_revision; r = git_revision; assert r == '%revision%', r + '!=%revision%'" || goto :error
    conda deactivate
    echo "--- Done! ---"
)

goto :EOF

:error
    echo Failed with error #%errorlevel%.
    exit /b %errorlevel%
```

</details>


All tests run successfully across all versions.

**Comment by peastman** [comment]
Thanks!

**Comment by jchodera** [comment]
Are we good to announce yet?

**Comment by peastman** [comment]
https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=12804&p=0&start=0&view=&sid=e9a3714fc54ff1f69803ec55611e5162

**Comment by jchodera** [comment]
@peastman : Going forward, is there some way we can migrate away from the SimTK forums for the primary release announcements? It would be much better if we could have something on GitHub---maybe a new issue we could point to?---where we could also edit the release announcement and collect feedback in a manner that is closely connected/linked to issues tagged here.

**Comment by peastman** [comment]
It's been a week since we posted the release candidate.  I've heard no feedback one way or another.  According to https://anaconda.org/conda-forge/openmm/files it has been downloaded 87 times: 58 on Linux, 11 on Mac, and 18 on Windows.

Should we move ahead with releasing it and creating builds for PDBFixer and OpenMM-Setup?

**Comment by jchodera** [comment]
As long as we have a rollback plan (to move it back to `test` if things go wrong) that we can execute once @jaimergp is on vacation, we should be good to release!

**Comment by peastman** [comment]
Let's go ahead then!  The OpenMM-Setup and PDBFixer releases should each be built from the head of the main branch.  That's cb256e2 for OpenMM-Setup and f02ecdb for PDBFixer.  I can create the Github releases.

We also need to put the up to date version of the documentation in the `latest` folder for docs.openmm.org.  I think you may need to do that.

**Comment by peastman** [comment]
Except that first I need to update the version number for OpenMM-Setup.  Just a moment.

**Comment by peastman** [comment]
Ok, the OpenMM-Setup revision is 015e21d.

**Comment by peastman** [comment]
Github releases have been created for all three repositories.

**Comment by jchodera** [comment]
> Let's go ahead then! The OpenMM-Setup and PDBFixer releases should each be built from the head of the main branch. That's cb256e2 for OpenMM-Setup and f02ecdb for PDBFixer. I can create the Github releases.

Can you remind me why we are building from commit hashes instead of releases? This is an unusual practice, and I do not understand why we are doing this.

**Comment by jaimergp** [comment]
PR to publish in `main` available here: https://github.com/conda-forge/openmm-feedstock/pull/41

I'll open the PRs for setup and pdbfixer at staged-recipes once that one is merged!

**Comment by jaimergp** [comment]
> Can you remind me why we are building from commit hashes instead of releases?

The releases are indeed based off tags now. The unusual cycle was during the RC builds. OpenMM, OpenMM-setup and PDBFixer are being built using tags now, but in the case of OpenMM we need the full repo to include the revision hash in the `version.py` file.

**Comment by jaimergp** [comment]
> As long as we have a rollback plan (to move it back to `test` if things go wrong)

I'll be around but in case this is needed urgently, that workaround is described [here](https://conda-forge.org/docs/orga/guidelines.html#fixing-broken-packages).

**Comment by jaimergp** [comment]
> PR to publish in `main` available here: [conda-forge/openmm-feedstock#41](https://github.com/conda-forge/openmm-feedstock/pull/41)

Ready to pull the trigger :)

**Comment by peastman** [comment]
Go for it!

**Comment by jaimergp** [comment]
Merged!

**Comment by jchodera** [comment]
@peastman: Can we note that we now deploy through conda-forge in the release notes, or at least link to the new installation instructions in the manual that specify installation via conda-forge?
https://github.com/openmm/openmm/releases/tag/7.5.0

**Comment by peastman** [comment]
As soon as we get the new documentation posted!  Are you planning to do that or should I?

**Comment by jchodera** [comment]
> As soon as we get the new documentation posted! Are you planning to do that or should I?

No idea what you're referring to, so could you go for it? 

We have (had?) a release checklist, but I'm not even sure where that lives or where we are on that process anymore.

**Comment by jchodera** [comment]
Let me know if you need any help from S3 side, though!

**Comment by peastman** [comment]
docs.openmm.org has a folder for each release containing its documentation, and another folder called `latest` that gets updated to always have the documentation for the latest release.  I'll update them.

**Comment by jchodera** [comment]
> docs.openmm.org has a folder for each release containing its documentation, and another folder called latest that gets updated to always have the documentation for the latest release. I'll update them.

Let me know if you need any changes from the S3 console.

**Comment by peastman** [comment]
Docs are uploaded and I added a link to the installation instructions.

**Comment by peastman** [comment]
How about the OpenMM-Setup and PDBFixer builds?  Are they ready to go?

**Comment by jaimergp** [comment]
Waiting for the CDN to sync up (+60 mins after seeing the package on anaconda.org). Once that happens (in 15 mins), I'll push to `staged-recipes` and then, after review, they will have their own feedstock and packages.

**Comment by jaimergp** [comment]
Before I submit the recipes, `pdbfixer` needs a version bump here: https://github.com/openmm/pdbfixer/blob/master/setup.py#L37

**Comment by peastman** [comment]
Done!  Sorry I missed that before.

**Comment by jaimergp** [comment]
I'll need the tag moved to that commit too, sorry! 😬

**Comment by peastman** [comment]
Done.  Some people are just so hard to please!  😀

**Comment by jchodera** [comment]
I've just tweeted the release announcement:
https://twitter.com/openmm_toolkit/status/1337118945640255488

I'll tweet more about the new 7.5.0 release over the next few days, but it would be great if we could update the benchmarks online.

**Comment by jaimergp** [comment]
PR for `openmm-setup` and `pdbfixer` available here: https://github.com/conda-forge/staged-recipes/pull/13443

Please add a comment to confirm you want to be a maintainer on the feedstocks. Thanks!

**Comment by jaimergp** [comment]
Both are live!

* OpenMM Setup: https://anaconda.org/conda-forge/openmm-setup/files
* PDBFixer: https://anaconda.org/conda-forge/pdbfixer/files

**Comment by peastman** [comment]
I think this means we can close the issue!  Thanks for the amazing work @jaimergp.
